ctpolicy/chromepolicy.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ctpolicy
+
+import (
+	"github.com/google/certificate-transparency-go/loglist"
+	"github.com/google/certificate-transparency-go/x509"
+)
+
+// googGroupBuilder is used to build log-group covering Google-operated logs.
+type googGroupBuilder struct {
+}
+
+func (g googGroupBuilder) Name() string {
+	return "Google-operated"
+}
+func (g googGroupBuilder) IsMember(log *loglist.Log) bool {
+	return log.GoogleOperated()
+}
+func (g googGroupBuilder) IsBase() bool {
+	return false
+}
+
+// nonGoogGroupBuilder is used to build log-group covering not-Google-operated logs.
+type nonGoogGroupBuilder struct {
+}
+
+func (g nonGoogGroupBuilder) Name() string {
+	return "Non-Google-operated"
+}
+func (g nonGoogGroupBuilder) IsMember(log *loglist.Log) bool {
+	return !log.GoogleOperated()
+}
+func (g nonGoogGroupBuilder) IsBase() bool {
+	return false
+}
+
+// ChromeCTPolicy provides policy logic in appliance to the data provided.
+type ChromeCTPolicy struct {
+	googBuilder  googGroupBuilder
+	nGoogBuilder nonGoogGroupBuilder
+	baseBuilder  baseGroupBuilder
+}
+
+// LogsByGroup describes submission requirements according to https://github.com/chromium/ct-policy/blob/master/ct_policy.md#qualifying-certificate.
+func (chromeP *ChromeCTPolicy) LogsByGroup(cert *x509.Certificate, approved *loglist.LogList) (map[string]*LogGroupInfo, error) {
+	var outerror error
+	groups := make(map[string]*LogGroupInfo)
+	googGroup := BuildGroup(chromeP.googBuilder, approved)
+	if err := googGroup.setMinInclusions(1); err != nil {
+		outerror = err
+	}
+	nonGoogGroup := BuildGroup(chromeP.nGoogBuilder, approved)
+	if err := nonGoogGroup.setMinInclusions(1); err != nil {
+		outerror = err
+	}
+
+	baseGroup := BuildGroup(chromeP.baseBuilder, approved)
+	switch m := lifetimeInMonths(cert); {
+	case m < 15:
+		if err := baseGroup.setMinInclusions(2); err != nil {
+			outerror = err
+		}
+	case m <= 27:
+		if err := baseGroup.setMinInclusions(3); err != nil {
+			outerror = err
+		}
+	case m <= 39:
+		if err := baseGroup.setMinInclusions(4); err != nil {
+			outerror = err
+		}
+	default:
+		if err := baseGroup.setMinInclusions(5); err != nil {
+			outerror = err
+		}
+	}
+	groups[googGroup.name] = &googGroup
```

Comment:
Could drop the `groups := make(...)` further up and just have:
```golang
groups := map[string]*LogGroupInfo{
	googGroup.name:    &googGroup,
	nonGoogGroup.name: &nonGoogGroup,
	baseGroup.name:	   &baseGroup,
}
```

Reply:
Dropped.